crates/matrix-sdk-crypto/src/machine/mod.rs

```diff
@@ -1336,7 +1359,54 @@ impl OlmMachine {
             e => self.handle_to_device_event(changes, &e).await,
         }
 
-        raw_event
+        Some(raw_event)
+    }
+
+    async fn event_is_from_dehydrated_device(
+        &self,
+        decrypted: &OlmDecryptionInfo,
+        sender_user_id: &UserId,
+    ) -> bool {
+        // Does the to-device message include device info?
+        if let Some(device_keys) = decrypted.result.event.sender_device_keys() {
+            // There is no need to check whether the device keys are signed correctly - any
+            // to-device message that claims to be from a dehydrated device is weird, so we
+            // will drop it.
+
+            // Does the included device info say the device is dehydrated?
+            if device_keys.dehydrated.unwrap_or(false) {
+                return true;
+            }
+            // If not, fall through and check our existing list of devices
+            // below, just in case the sender is sending us
+            // incorrect information embedded in the to-device message, but we
+            // know better.
+        }
+
+        // Do we already know about this device?
```

Comment:
My tests would still pass if the rest of this function did not exist.

I looked into how to construct a to-device message that was from a known dehydrated device, but did not itself contain `device_keys` and it just looked too complex. I even got part-way in to an implementation pairing with Rich, but it just seemed so much!